mobile/src/main/java/rs/readahead/washington/mobile/util/jobs/TellaUploadJob.java

```diff
@@ -0,0 +1,158 @@
+        if (key == null) { // key disappeared
+            return exit(Job.Result.RESCHEDULE);
+        }
+
+        dataSource = DataSource.getInstance(getContext(), key);
+        List<MediaFile> mediaFiles = dataSource.getUploadMediaFiles(ITellaUploadsRepository.UploadStatus.SCHEDULED).blockingGet();
+
+        if (mediaFiles.size() == 0) {
+            return exit(Job.Result.SUCCESS);
+        }
+
+        TellaUploadServer server = dataSource.getTellaUploadServer(Preferences.getAutoUploadServerId()).blockingGet();
+        if (server == TellaUploadServer.NONE) {
+            return exit(Result.FAILURE);
+        }
+        final TUSClient tusClient = new TUSClient(getContext(), server.getUrl(), server.getUsername(), server.getPassword());
+
+        for (MediaFile file : mediaFiles) {
+            fileMap.put(String.valueOf(file.getFileName()), file);
+        }
+
+        Flowable.fromIterable(mediaFiles)
+                .flatMap(tusClient::upload)
+                .blockingSubscribe(this::updateProgress, throwable -> {
+                    if (throwable instanceof NoConnectivityException) {
+                        exitResult = Result.RESCHEDULE;
+                        return;
+                    }
+                    Timber.d(throwable);
+                    Crashlytics.logException(throwable);
+                });
+
+        if (exitResult != null) {
+            exit(exitResult);
+        }
+
+        return exit(Job.Result.SUCCESS);
+    }
+
+    @Override
+    protected void onReschedule(int newJobId) {
+    }
+
+    private boolean enter() {
+        synchronized (TellaUploadJob.class) {
+            boolean current = running;
+            running = true;
+            return !current;
+        }
+    }
+
+    private Job.Result exit(Job.Result result) {
+        synchronized (TellaUploadJob.class) {
+            running = false;
+            return result;
+        }
+    }
+
+    public static void scheduleJob() {
+        new JobRequest.Builder(TellaUploadJob.TAG)
+                .setExecutionWindow(1_000L, 10_000L) // start between 1-10sec from now
+                .setBackoffCriteria(10_000L, JobRequest.BackoffPolicy.LINEAR)
+                .setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
+                .setRequirementsEnforced(true)
+                //.setPersisted(true) // android.Manifest.permission.RECEIVE_BOOT_COMPLETED
+                .setUpdateCurrent(true) // also, jobs will sync them self while running
+                .build()
+                .schedule();
+    }
+
+    private void updateProgress(UploadProgressInfo progressInfo) {
+
+        if (fileMap.get(progressInfo.name) == null) return;
+
+        MediaFile mediaFile = fileMap.get(progressInfo.name);
+        assert mediaFile != null;
+
+        switch (progressInfo.status) {
+
+            case STARTED:
+                dataSource.setUploadStatus(mediaFile.getId(), ITellaUploadsRepository.UploadStatus.UPLOADING).blockingAwait();
```

Comment:
As `Flowable` in TUSClient use `BackpressureStrategy.LATEST`, any of these events can be overridden, and upload process should handle that. In this case, `STARTED` and `OK` cases can be handled equally with single `DataSource` method, setting status to `UPLOADING` and updating uploaded size.